### PR TITLE
feat: add basic C++ support

### LINKED ISSUE #953: Add Support for CPP language
#### Reason/Context

I have JSON objects that are consumed/produced by Java and CPP (c++) programs. I would love to be able to generate the model for both of them with a single tool. 

 #### Description

Implement a CPP (c++) class generator

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds a basic C++ generator.

**Remaining tasks**
This can either be done after or now.
- [ ] Add enum key to value support function for each enum
- [ ] Add serialization support

**Related issue(s)**
Fixes #953

### CODE REVIEW

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusConstrainer.ts:18]: IIUC the Object should be different from Reference in that it owns the property / field data, instead of referencing another object. In that case the `*` for pointer type should not be here, and the property's type definition must be visible (by including that type's header).

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusConstrainer.ts:20]: Here there is the additional question of whether or not the reference allows the user to modify the referenced value. If modifications are not intended, we should add the `const` keyword to before the `*`.

File: src/generators/cplusplus/CplusplusFileGenerator.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusFileGenerator.ts:35]: It would make more sense to make the generated files headers, with *.hpp extension.

Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusFileGenerator.ts:35]: So we only generate header files, no `.cpp` files? 🤔

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusConstrainer.ts:20]: I dont see how we can force no modifications here 🤔 

We could potentially add this as an option later down the line if it gets requested?

File: src/generators/cplusplus/CplusplusFileGenerator.ts
Comment (beku-epitome) [reply src/generators/cplusplus/CplusplusFileGenerator.ts:35]: Yes, we are basically creating an interface for data storage, and interfaces need to be shared between compilation units, that's what headers achieve.

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [reply src/generators/cplusplus/CplusplusConstrainer.ts:20]: If the output were `${constrainedModel.name} const*`, that would mean that whoever is accessing this property, cannot modify it. My question was more about the theoretical model than the C++ output though, how these References are intended to behave.

Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusConstrainer.ts:20]: Is that not up to the individual use-case, i.e. not really something we can hardcode but should be conditional based on what the user want?

File: src/generators/cplusplus/CplusplusFileGenerator.ts
Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusFileGenerator.ts:35]: Okay, so we can change all the models to be located in header files instead?

Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusFileGenerator.ts]: Alright, switched 👍

File: examples/generate-cplusplus-models/README.md
Comment (beku-epitome) [inline examples/generate-cplusplus-models/README.md:3]: Copy paste error, need to replace TypeScript to C++

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusConstrainer.ts:23]: I recently learned about [std::any](https://en.cppreference.com/w/cpp/utility/any), let's use that here instead of `void*`. It needs a `'#include <utility>'` dependency.

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusConstrainer.ts:51]: It doesn't make sense to use indirection here, store the enum by value instead. The `*` should be removed.

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [reply src/generators/cplusplus/CplusplusConstrainer.ts:20]: Now I finally understand this, and the desired solution is to have only `${constrainedModel.name}`, without the `*`. I was initially confused by the naming, but here we are creating a reference to a type, not to an object. It is important to note that the referenced type must be defined before the current type is defined.

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusConstrainer.ts:23]: Okay 👍

File: src/generators/cplusplus/Constants.ts
Comment (jonaslagoni) [reply src/generators/cplusplus/Constants.ts:127]: Alright, removed those you highlighted here.

**Comment by jonaslagoni** [comment]
We have a remaining problem when the struct contains properties with references, it seems that in those cases we have to instead of 
```
std::variant<core_space_schema_space_meta_minus_schema_object, bool> additional_items;
```
do the following:
```
std::variant<core_space_schema_space_meta_minus_schema_object, bool> *additional_items;
```

Does that make sense to you @beku-epitome?

**Comment by beku-epitome** [comment]
> We have a remaining problem when the struct contains properties with references, it seems that in those cases we have to instead of
> 
> ```
> std::variant<core_space_schema_space_meta_minus_schema_object, bool> additional_items;
> ```
> 
> do the following:
> 
> ```
> std::variant<core_space_schema_space_meta_minus_schema_object, bool> *additional_items;
> ```
> 
> Does that make sense @beku-epitome?

No, this is not what we should do, as this way we'd need to do dynamic memory management for trivially copyable types. The proper solution is to add the reference type into the dependencies, so the header file where the reference type is defined gets included into the current header. If there is only one file generated, then the reference type definition must stand before any occurrences of references to it. This also mandates that there is no circular referencing.

**Comment by jonaslagoni** [comment]
Okay we are down to the last problem I think 🤔 

Currently a property is often named the same as the referenced model i.e. look at `category`

```
```

**Comment by beku-epitome** [comment]
@jonaslagoni I believe if you fully qualify the type's namespace, the problem should go away:
```
namespace TestNamespace{
  struct pet_post_body {
    int id;
    TestNamespace::category category;
    std::string name;
    std::vector<std::variant<std::string, std::any>> photo_urls;
    std::vector<std::variant<TestNamespace::tags_item, std::any>> tags;
    TestNamespace::status status;
    std::map<std::string, std::any> additional_properties;
  };
}
```

**Comment by jonaslagoni** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=asyncapi_generator-model-sdk&pullRequest=1156)
> 
> [Bug](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=SECURITY_HOTSPOT) [E](https://sonarcloud.io/project/security_hotspots?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=SECURITY_HOTSPOT) [2 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=CODE_SMELL) [4 Code Smells](https://sonarcloud.io/project/issues?id=asyncapi_generator-model-sdk&pullRequest=1156&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=asyncapi_generator-model-sdk&pullRequest=1156) No Coverage information [](https://sonarcloud.io/component_measures?id=asyncapi_generator-model-sdk&pullRequest=1156&metric=new_duplicated_lines_density&view=list) [0.1% Duplication](https://sonarcloud.io/component_measures?id=asyncapi_generator-model-sdk&pullRequest=1156&metric=new_duplicated_lines_density&view=list)

I am fine with these errors 👍

**Comment by jonaslagoni** [comment]
> @jonaslagoni I believe if you fully qualify the type's namespace, the problem should go away:

@beku-epitome worked 👌

**Comment by jonaslagoni** [comment]
hey @beku-epitome just pinging you in-case you missed it ✌️

**Comment by beku-epitome** [comment]
Sorry for the delay, I already wrote a response, but my session got messed up in that window, so it wasn't sent :sweat_smile: Here's what I wanted to say:
Great job so far! My notes on the current output:
1. `additional_properties` field shouldn't be added by default, only if the model contains this flag explicitly.
2. Don't add `std::any` variant to arrays, only have the explicitly defined types in the array, and if only one type is defined for the array, don't use variant at all, just have `std::vector<int> array;` in the simple case.
3. It would be nice to code into C++ which properties are required, and which are optional. The current output tells that everything is required. To make properties optional, their type should be wrapped in `std::optional<int>` (`#include <optional>` dependency is added).

**Comment by jonaslagoni** [comment]
>Sorry for the delay, I already wrote a response, but my session got messed up in that window, so it wasn't sent 😅 Here's what I wanted to say:

No worries, thanks for the review§

> additional_properties field shouldn't be added by default, only if the model contains this flag explicitly.

This is the default behavior of the input that affects this, not the generator, it just does what it's told from the inputs perspective, because in AsyncAPI and JSON Schema draft 7 `additionalProperties: true` it renders the property, if your input is set to having `additionalProperties: false`, the property will no be there 🙂 

> Don't add std::any variant to arrays, only have the explicitly defined types in the array, and if only one type is defined for the array, don't use variant at all, just have std::vector<int> array; in the simple case.

I don't think I fully understand this, can you clarify this?

> It would be nice to code into C++ which properties are required, and which are optional. The current output tells that everything is required. To make properties optional, their type should be wrapped in std::optional<int> (#include <optional> dependency is added). 

This should definitely be part of the default yea, gonna adapt the generator 🙂

**Comment by beku-epitome** [comment]
Regarding the array output, if the input is an array property with multiple types, e.g.
```
{
    "type": "array",
    "items":
    [
        {
            "type": "string"
        },
        {
            "type": "number"
        }
    ]
}
```
then the current output is `std::vector<std::variant<std::string, double, std::any>> array_type;`. I'd like the output to be `std::vector<std::variant<std::string, double>> array_type;` instead.
If the input property is like this:
```
{
    "type": "array",
    "items":
    [
        {
            "type": "number"
        }
    ]
}
```
then the current output is `std::vector<std::variant<double, std::any>> array_type;`. I'd like the output to be `std::vector<double> array_type;` in this case.

**Comment by jonaslagoni** [comment]
@beku-epitome thats a problem with the underlying interpreting functionality, it's fixed as part of `v2`, nothing we need to do on our side 🙂

**Comment by jonaslagoni** [comment]
@beku-epitome alright, adapted the types to include optional ones, please take a look 🙂

File: docs/usage.md
Comment (beku-epitome) [inline docs/usage.md:205]: Typo: `Kotlin` -> `C++` or `modern C++`

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (beku-epitome) [inline src/generators/cplusplus/CplusplusConstrainer.ts:22]: This doesn't seem to do what I expect it to do. Even if I mark an object's property as required, it still ends up being generated as an optional field.

**Review by beku-epitome — CHANGES_REQUESTED** [review]
I have only two notes, see inline, otherwise looks good :)

File: src/generators/cplusplus/CplusplusConstrainer.ts
Comment (jonaslagoni) [reply src/generators/cplusplus/CplusplusConstrainer.ts]: A bit too fast with the implementation 😅 Was missing `&& partOfProperty.required !== true`.

✅

**Comment by jonaslagoni** [comment]
✅ @beku-epitome

**Review by beku-epitome — APPROVED** [review]
If the whole array trouble is resolved somewhere else, then I'm happy with the result :)

**Comment by jonaslagoni** [comment]
@beku-epitome would you like to become a champion for the c++ generator? You dont necessarily need to code, but it can evolve around issues, discussions and reviews of PRs? 🙂

We have some more information about championing in the docs.

**Comment by jonaslagoni** [comment]
@magicmatatjahu @kennethaasan need a review/approval 👍

**Comment by beku-epitome** [comment]
> @beku-epitome would you like to become a champion for the c++ generator? You dont necessarily need to code, but it can evolve around issues, discussions and reviews of PRs? slightly_smiling_face
> 
> We have some more information about championing in the docs.

If it's a similar level of involvement that was required here I can be of service :)

**Review by kennethaasan — APPROVED** [review]

**Comment by jonaslagoni** [comment]
@beku-epitome thanks for all the reviews, really appreciate your time 🙏 

@kennethaasan thanks for the review 💪 

> If it's a similar level of involvement that was required here I can be of service :)

@beku-epitome yep more or less, you can lead the C++ development how you see fit, if a PR targets C++ generator, you can review PRs all on your own 🙂 

If you have ways you want the generator improved, a good way is to create issues for them in a good first issue type of approach, enabling others to pick them up. That way you can lead it without touching typescript 😄 So yea, basically championing the generator 🙂 

If you want to, just add yourself to the [codeowners](https://github.com/asyncapi/modelina/blob/cbb240bdbf7caedc0bc6abc5465946428c63bad5/CODEOWNERS#L49) file for C++, afterwards I'll invite you to the team, and if you want, you can join the AsyncAPI TSC 🙂

**Comment by jonaslagoni** [comment]
@all-contributors please add @beku-epitome for code, review, test

**Pull request merged**
